File: Problem-19/BucketSort.py
# WAP to sort an Array using Bucket sort.  

def bucketSort(arr) :
    n = len(arr)
    Max = max(arr)

    buckets = [[] for _ in range(n)]

    for i in arr :
        index = (i * n) // (Max + 1)
        buckets[index].append(i)

    for i in buckets :
        i.sort()

    sorted_arr = []
    for i in buckets :
        sorted_arr.extend(i)

    return sorted_arr


n = int(input("Enter n : "))

arr = []

print("Enter elements into arr : ")
for i in range(n) :
    element = int(input(f"arr[{i}] : "))
    arr.append(element)


print("\nUnsorted Array : ", end="\n")
for i in arr :
    print(i, end="  ")


arr = bucketSort(arr)


print("\nSorted Array : ", end="\n")
for i in arr :
    print(i, end="  ")

File: Problem-19/BucketSort.c
// WAP to sort an Array using Bucket sort.  

#include <stdio.h>

void insertionSort(int arr[], int n){
    for(int i = 1; i < n; i++){
        int key = arr[i];
        int j = i - 1;

        while(j >= 0 && arr[j] > key){
            arr[j + 1] = arr[j];
            j--;
        }

        arr[j + 1] = key;
    }
}

void main(){
    int n;

    printf("Enter n : ");
    scanf("%d", &n);

    int arr[n];

    printf("Enter elements into arr : \n");
    for(int i = 0; i < n; i++){
        printf("arr[%d] : ", i);
        scanf("%d", &arr[i]);
    }

    printf("\nUnsorted Array : ");
    for(int i = 0; i < n; i++){
        printf("%d  ", arr[i]);
    }

    int max = arr[0];
    for(int i = 1; i < n; i++){
        if(arr[i] > max){
            max = arr[i];
        }
    }

    int buckets[n][n];
    int bucketCount[n];

    for(int i = 0; i < n; i++){
        bucketCount[i] = 0;
    }

    for(int i = 0; i < n; i++){
        int index = (arr[i] * n) / (max + 1);
        buckets[index][bucketCount[index]++] = arr[i];
    }

    for(int i = 0; i < n; i++){
        if(bucketCount[i] > 0){
            insertionSort(buckets[i], bucketCount[i]);
        }
    }

    int k = 0;
    for(int i = 0; i < n; i++){
        for(int j = 0; j < bucketCount[i]; j++){
            arr[k++] = buckets[i][j];
        }
    }

    printf("\nSorted Array : ");
    for(int i = 0; i < n; i++){
        printf("%d  ", arr[i]);
    }

}
